src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,104 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final double GROW_MULTIPLY = 0.5;
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private T[] array;
+
+    public ArrayList() {
+        array = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        growCheck();
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        growCheck();
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Element not found at " + index);
+        }
+        System.arraycopy(array, index, array, index + 1, size - index);
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isNotValidValue(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Element does not exist");
+        }
+        return (T)array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isNotValidValue(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Element does not exist");
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isNotValidValue(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Element does not exist in position"
+                    + index);
+        }
+        T tempVal = array[index];
+        System.arraycopy(array, index + 1, array, index, --size - index);
+        return tempVal;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (array[i] == element
+                    || array[i] != null && array[i].equals(element)) {
+                return remove(i);
+            }
+        }
+
+        throw new NoSuchElementException("Element does not exist");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void growCheck() {
+        if (size == array.length) {
+            grow();
+        }
+    }
+
+    private void grow() {
+        int newCapacity = size + (int)(DEFAULT_CAPACITY * GROW_MULTIPLY);
```

Comment:
This is not how the array length is expected to be increased. You should multiply an array length should be multiplied by 1.5 each time. But here, your array length is increased by 5 units only.